src/development/vortex/development/core/pipelines/training_pipeline.py

```diff
@@ -317,45 +322,38 @@ def run(self, save_model: bool = True) -> EasyDict:
                                 "result, choose either one of [{}]".format(metric_name, val_res_key))
-                        if best_metric[metric_name] < val_results[metric_name]:
-                            best_metric[metric_name] = val_results[metric_name]
+                        if self.best_metrics[metric_name] < val_results[metric_name]:
+                            self.best_metrics[metric_name] = val_results[metric_name]
                             model_fname = "{}-best-{}.pth".format(self.config.experiment_name, metric_name)
                             self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
 
                 print('epoch %s validation : %s' % (epoch, ', '.join(['{}:{:.4f}'.format(key, value) for key, value in val_results.items()])))
 
             ## save on best loss
-            if self.save_best_type and 'loss' in self.save_best_type and best_loss > loss.item() and save_model:
-                best_loss = loss.item()
-                metrics = None
-                if self.valid_for_validation:
-                    metrics = val_results
+            if self.save_best_type and 'loss' in self.save_best_type and \
+                    self.best_metrics['loss'] > loss.item() and save_model:
+                self.best_metrics['loss'] = loss.item()
                 model_fname = "{}-best-loss.pth".format(self.config.experiment_name)
-                self._save_checkpoint(epoch, metrics=metrics, filename=model_fname)
+                self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
 
             # Save on several epoch
             if self.save_epoch and ((epoch+1) % self.save_epoch == 0) and save_model:
-                metrics = None
-                if self.valid_for_validation:
-                    metrics = val_results
                 model_fname = "{}-epoch-{}.pth".format(self.config.experiment_name, epoch)
-                self._save_checkpoint(epoch, metrics=metrics, filename=model_fname)
-
-        # Save final weights on after all epochs finished
-        if save_model:
-            metrics = None
-            if self.valid_for_validation:
-                metrics = val_results
-            model_fname = "{}.pth".format(self.config.experiment_name)
-            model_path = self._save_checkpoint(last_epoch, metrics=metrics, filename=model_fname)
-            # Copy final weight from runs directory to experiment directory
-            shutil.copy(model_path, Path(self.experiment_directory))
+                self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
+
+            if save_model: ## save this epoch
+                model_fname = "{}-last.pth".format(self.config.experiment_name)
+                model_path = self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
+                if epoch == self.config.trainer.epoch-1:
```

Comment:
please rename the final epoch with the same name which is copied to experiment directory

Reply:
why the same name? you mean with `-last.pth` suffix?

Reply:
nope, the final weight should not have -last.pth suffix thats why in the run directory the -last suffix need to be renamed. In this code, you only rename it when it's copied to experiment directory

Reply:
ahh, so in the run directory also needs to be renamed? is it copied or simply renamed? wouldn't it be redundant if it is copied?

Reply:
yes, it needs to be redundant for backup purpose in the case of final weight. I think if we have final weight we didnt need the `-last` suffix so it's enough to be renamed then copied to experiment dir

Reply:
then if it is renamed, why not just have the filename without `-last.pth` suffix?

Reply:
I mean why not just have it without `-last.pth` suffix from the beginning? isn't it going to be confusing if the weight with the `-last.pth` suffix is gone in the end of the training?

Reply:
then the user will be confused why do they have final weight if the training is not finished. Since our first release we state that final weight is exist in run and experiment dir after finish training. It will be more confusing if suddenly after update they have different name for final weight in the run directory

Reply:
We need to think about past behavior and not abruptly change it for new features, adding last checkpoint will helpful. However if the training is finished why do we need last checkpoint though? thats my argument for renaming it when the training complete

Reply:
well the last checkpoint is basically the final weight though. okay then

Reply:
resolved in b17525e